src/utils.c

```diff
@@ -1025,6 +1025,266 @@ ts_relation_size_impl(Oid relid)
+ */
+static int64
+ts_try_relation_cached_size(Relation rel, bool verbose)
+{
+	BlockNumber result = 0, nblocks = 0;
+	ForkNumber forkNum;
+	bool cached = true;
+
+	/* Get heap size, including FSM and VM */
+	for (forkNum = 0; forkNum <= MAX_FORKNUM; forkNum++)
+	{
+#if PG14_LT
+		/* PG13 does not have smgr_cached_nblocks */
+		result = InvalidBlockNumber;
+#else
+		result = RelationGetSmgr(rel)->smgr_cached_nblocks[forkNum];
+#endif
+
+		if (result != InvalidBlockNumber)
+		{
+			nblocks += result;
+		}
+		else
+		{
+			if (smgrexists(RelationGetSmgr(rel), forkNum))
+			{
+				cached = false;
+				nblocks += smgrnblocks(RelationGetSmgr(rel), forkNum);
+			}
+		}
+	}
+
+	if (verbose)
+		ereport(DEBUG2,
+				(errmsg("%s for %s",
+						cached ? "Cached size used" : "Fetching actual size",
+						RelationGetRelationName(rel))));
+
+	/* convert the size into bytes and return */
+	return nblocks * BLCKSZ;
+}
+
+static RelationSize
+ts_relation_approximate_size_impl(Oid relid)
+{
+	RelationSize relsize = { 0 };
+	Relation rel;
+
+	DEBUG_WAITPOINT("relation_approximate_size_before_lock");
+	/* Open relation earlier to keep a lock during all function calls */
+	rel = try_relation_open(relid, AccessShareLock);
+
+	if (!rel)
+		return relsize;
+
+	/* Get the main heap size */
+	relsize.heap_size = ts_try_relation_cached_size(rel, true);
+
+	/* Get the size of the relation's indexes */
+	if (rel->rd_rel->relhasindex)
+	{
+		List *index_oids = RelationGetIndexList(rel);
+		ListCell *cell;
+
+		foreach (cell, index_oids)
+		{
+			Oid idxOid = lfirst_oid(cell);
+			Relation idxRel;
+
+			idxRel = relation_open(idxOid, AccessShareLock);
+			relsize.index_size += ts_try_relation_cached_size(idxRel, false);
+			relation_close(idxRel, AccessShareLock);
+		}
+	}
+
+	/* If there's an associated TOAST table, calculate the total size (including its indexes) */
+	if (OidIsValid(rel->rd_rel->reltoastrelid))
+	{
+		Relation toastRel;
+		List *index_oids;
+		ListCell *cell;
+
+		toastRel = relation_open(rel->rd_rel->reltoastrelid, AccessShareLock);
+		relsize.toast_size = ts_try_relation_cached_size(toastRel, false);
+
+		/* Get the indexes size of the TOAST relation */
+		index_oids = RelationGetIndexList(rel);
+		foreach (cell, index_oids)
+		{
+			Oid idxOid = lfirst_oid(cell);
+			Relation idxRel;
+
+			idxRel = relation_open(idxOid, AccessShareLock);
+			relsize.toast_size += ts_try_relation_cached_size(idxRel, false);
+			relation_close(idxRel, AccessShareLock);
+		}
+
+		relation_close(toastRel, AccessShareLock);
+	}
+
+	relation_close(rel, AccessShareLock);
+
+	/* Add up the total size based on the heap size, indexes and toast */
+	relsize.total_size = relsize.heap_size + relsize.index_size + relsize.toast_size;
+
+	return relsize;
+}
+
+TS_FUNCTION_INFO_V1(ts_relation_approximate_size);
+Datum
+ts_relation_approximate_size(PG_FUNCTION_ARGS)
+{
+	Oid relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
```

Comment:
```suggestion
	Oid relid = PG_GETARG_OID(0);
```
If you mark this function as STRICT you don't need this check.

Reply:
done.